### ISSUE TITLE
Estimator built with keras.estimator.model_to_estimator fails on Estimator.export_savedmodel

### ISSUE THREAD

**Issue by zmjjmz** [tensorflow/tensorflow#16385]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: Yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Debian 3.16.36
- **TensorFlow installed from (source or binary)**: Binary
- **TensorFlow version (use command below)**: ('v1.4.0-19-ga52c8d9', '1.4.1')
- **Python version**: 2.7.9
- **Bazel version (if compiling from source)**:
- **GCC/Compiler version (if compiling from source)**:
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: See gist

### Describe the problem

If I create a model with `tf.keras`, compile it and then turn it into an estimator by simply passing it thru to `tf.keras.estimator.model_to_estimator`, I am able to train and evaluate the model just fine -- however when I got to export it with `Estimator.export_savedmodel`, I get the following error:

```
Traceback (most recent call last):
  File "endtoend_noweights_trainer_keras.py", line 302, in <module>
    get_serving_input_fn(hyperparameters),
  File "/home/u1/zach/proj/dataplayground2/local/lib/python2.7/site-packages/tensorflow/python/estimator/estimator.py", line 517, in export_savedmodel
    serving_input_receiver.receiver_tensors_alternatives)
  File "/home/u1/zach/proj/dataplayground2/local/lib/python2.7/site-packages/tensorflow/python/estimator/export/export.py", line 193, in build_all_signature_defs
    raise ValueError('export_outputs must be a dict.')
ValueError: export_outputs must be a dict.
```

I'm not sure what `export_outputs` is, but if I had to guess it should be a mapping of output names to output tensors from the `keras` model.

Here's the (very unclean) [code](https://gist.github.com/zmjjmz/8e3a7e5430f2e700a9e89bf2b4f6259b) I'm using to get to this, although there's a lot of dependencies that won't work for y'all. If you need a repro I can take the time to put it together, just let me know. Notably the error occurs on line `300`.

**Comment by tatatodd** [comment]
@zmjjmz yes, please create a minimum repro.

Adding @fchollet in case there's anything obvious that might be wrong.

**Comment by zmjjmz** [comment]
Ok, sure, here it is: https://gist.github.com/zmjjmz/9b013c6524839512b85dfc08a56f28c2

It's possible that something about how I defined `serving_input_fn` is incorrect (it works for normal estimators), but I also couldn't find any documentation on how to properly / differently define it for an Estimator that came from `tf.keras.estimator.model_to_estimator`.

**Comment by tatatodd** [comment]
I'm not a python or keras expert, but your imports are not idiomatic:

You're doing this:
```
import tensorflow
from tensorflow.python.keras._impl import keras
from tensorflow.python.estimator.export.export_output import PredictOutput
from tensorflow.python.estimator.export.export import build_raw_serving_input_receiver_fn
from tensorflow.python.saved_model import signature_constants
from tensorflow.python import debug as tf_debug
```

Normally we do this:
```
import tensorflow as tf
```

Then in your code, you would refer to things as, e.g. `tf.keras.estimator` and things like that.  I don't know whether this has anything to do with your actual issue, but bypassing the standard `__init__.py` logic seems like a bad idea.  Perhaps you can change that first?

**Comment by zmjjmz** [comment]
It's certainly unidiomatic -- I probably did that because in the original code this repro comes from I was converting some code (with a ton of custom layers) that was used to the `keras` structure and `tf.keras` is slightly different (e.g. `tf.keras.layers.core.Dense` doesn't exist, but is instead `tf.keras.layers.Dense`). 

I'll rectify that and see if I can reproduce, but I'd be surprised (and still concerned) if that's the issue.

**Comment by fchollet** [comment]
Could you share a short, standalone code snippet to reproduce this issue?

**Comment by zmjjmz** [comment]
I believe [this](https://gist.github.com/zmjjmz/1ebd3f6b78f012326a660ad126cdd4d5) is standalone and hopefully isn't too long.

**Comment by yazsh** [comment]
I'm also having this exact problem.

**Comment by random-residual** [comment]
I am having this problem as well on AWS SageMaker. I am using a simple example and transforming it to an estimator with ```python tf.keras.estimator.model_to_estimator(keras_model=model)```. I can train the model but cannot save it, I get the following error:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-27-ed0372b11bae> in <module>()
      1 
      2 
----> 3 exported_model = model.export_savedmodel(export_dir_base = 'export/Servo/', serving_input_receiver_fn = serving_input_fn)
      4 
      5 print (exported_model)

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/tensorflow/python/estimator/estimator.py in export_savedmodel(self, export_dir_base, serving_input_receiver_fn, assets_extra, as_text, checkpoint_path)
    515           serving_input_receiver.receiver_tensors,
    516           estimator_spec.export_outputs,
--> 517           serving_input_receiver.receiver_tensors_alternatives)
    518 
    519       if not checkpoint_path:

~/anaconda3/envs/tensorflow_p36/lib/python3.6/site-packages/tensorflow/python/estimator/export/export.py in build_all_signature_defs(receiver_tensors, export_outputs, receiver_tensors_alternatives)
    191     receiver_tensors = {_SINGLE_RECEIVER_DEFAULT_NAME: receiver_tensors}
    192   if export_outputs is None or not isinstance(export_outputs, dict):
--> 193     raise ValueError('export_outputs must be a dict.')
    194 
    195   signature_def_map = {}

ValueError: export_outputs must be a dict.

```

My code is:

```python


import numpy as np
import os
import tensorflow as tf
from sklearn.preprocessing import LabelEncoder
from sklearn.externals import joblib


def featureTransform(features, max_words):
    tokenize = tf.keras.preprocessing.text.Tokenizer(num_words=max_words, char_level=False)
    tokenize.fit_on_texts(features) 
    return tokenize.texts_to_matrix(features).astype(np.float32) 

def encodeLabels(labels):
    encoder = LabelEncoder()
    encoder.fit(labels)
    y = encoder.transform(labels)
    num_classes = np.max(y) + 1
    print("num classes: {}".format(num_classes))
    return tf.keras.utils.to_categorical(y, num_classes).astype(np.float32)

def estimator_fn(run_config, params):
    model = tf.keras.models.Sequential()
    model.add(tf.keras.layers.Dense(500, activation='relu', input_shape=(500,), name="features"))
    model.add(tf.keras.layers.Dropout(0.5))
    model.add(tf.keras.layers.Dense(500, activation='relu'))
    model.add(tf.keras.layers.Dropout(0.5))
    model.add(tf.keras.layers.Dense(123, activation='softmax'))
    model.compile(loss='categorical_crossentropy',
              optimizer='rmsprop',
              metrics=['accuracy'])
    print("input names: {}".format(model.input_names))
    print("output names: {}".format(model.output_names))
    
    return tf.keras.estimator.model_to_estimator(keras_model=model)

def serving_input_fn():
    feature_spec = {'features_input': tf.FixedLenFeature(dtype=tf.float32, shape=[500])}
    return tf.estimator.export.build_parsing_serving_input_receiver_fn(feature_spec)()

def train_input_fn(training_dir, params):
    """Returns input function that would feed the model during training"""
    return _generate_input_fn(training_dir, 'assignment_train.csv')

def eval_input_fn(training_dir, params):
    """Returns input function that would feed the model during evaluation"""
    return _generate_input_fn(training_dir, 'assignment_test.csv')


def _generate_input_fn(training_dir, training_filename, shuffle=False):
    training_set = tf.contrib.learn.datasets.base.load_csv_without_header(
    filename=os.path.join(training_dir, training_filename), target_dtype=np.str, features_dtype=np.float32)
    
    input_fn = tf.estimator.inputs.numpy_input_fn(
        x={"features_input":  np.array(training_set.data)}, 
        y=encodeLabels(training_set.target),
        num_epochs=100,
        shuffle=shuffle
    )
    return input_fn

```

I can train the model locally in Jupyter in SageMaker.

```python

from assignments import estimator_fn, _generate_input_fn
from assignments import serving_input_fn

model = estimator_fn(run_config = None, params = None)

model.train(input_fn=_generate_input_fn('data','assignment_train.csv', shuffle=True))


exported_model = model.export_savedmodel(export_dir_base = 'export/Servo/', serving_input_receiver_fn = serving_input_fn)

print (exported_model)
import tarfile
with tarfile.open('model.tar.gz', mode='w:gz') as archive:
    archive.add('export', recursive=True)

```

I have spend quite a bit of time trying to get around this.

**Comment by yifeif** [comment]
Which version of TensorFlow were you using? Do you still see this issue with version 1.5+?

**Comment by random-residual** [comment]
It seems that SageMaker's TensorFlow version is '1.4.0'. I suppose that means I am stuck with this until they update the version?

I ended up writing a custom estimator in TensorFlow with Keras layers, i.e.:

```python first_hidden_layer = tf.keras.layers.Dense(128, activation='relu', name='first-layer')(features[INPUT_TENSOR_NAME])```

However I noticed that when I trained on SageMaker's infrastructure it would give me a dimension mismatch if my labels were one-hot encoded, I have to just use label encoding. I think this follows how the pre-made TensorFlow estimators work. However I had no problem using one-hot encoding in Keras to train locally, but did have a problem locally training the estimator with Keras with encoded labels:

```Tensor conversion requested dtype int64 for Tensor with dtype float32```

Currently I am going to stick to the custom estimators. Thanks for responding.

**Comment by yifeif** [comment]
Tried the repro snippets with TF 1.5 and I don't see the error anymore. Let me know if this is still an issue.
Btw @random-residual the type cast issue you mentioned should be fixed with [this commit](https://github.com/tensorflow/tensorflow/commit/4c86ece040cb96ea689f5c0d084b6959274eab91). It should be available in master and v1.6.0-rc0+.

**Comment by yifeif** [comment]
Btw took another look and the export_savedmodel issue should have been fixed in [this commit](https://github.com/tensorflow/tensorflow/commit/fe8406149feec453250905965a14285465cd2063), which is available in v.1.5.0-rc0+. Closing this issue. Thanks!
